### PR TITLE
Improvements to Emails in API v2 [PLAT-1125] [PLAT-1150]

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose

Many API V2 email improvements rolled into one!

### confirmation/verification
Adding an email involves a 2 step process: clicking a confirmation link in an email sent to the address you wish to confirm, and then also confirming a second time in a modal on the OSF side. The `Email` object isn't actually created until that modal is clicked, and the email remains in a middle ground til that happens.

To capture that middle ground, we added a `verified` property. When an unconfirmed email is first added, it has both `confirmed=False` and `verified=False`. If an email has been confirmed via email, that API entry will have `confirmed=True` but still `verified=False`.

When the email is removed from the user's `email_verifications`, when `user.confirm_email()` is called, then both `confirmed=True` and `verified=True`, and the `Email` object will have been created.

### filter by primary email
It'd be nice to be able to filter by primary/non-primary, so now do that with `filter[primary]=True` or `False`!

### re-send email verifications with `GET` request and query param
API v1 provided a way to re-send confirmation emails, so let's do that here, too! Send a `GET` request to an email's detail endpoint with the query param `?resend_confirmation=True` to re-send the confirmation email.
In addition, if the email detail is an unconfirmed email, it will have a key in the `links` dict pointing to the URL to use to resend a confirmation email.

## Changes

- Add filtering to `UserEmailsList`
- Add `verified` field
- Allow `verified` to be updated via an API request to a `confirmed` email address
- Tests for updating, filtering, values for `created` and `verified`
- Add filtering by `primary` field
- Send a `GET` request to an email's detail endpoint with the query param `?resend_confirmation=True` to re-send the confirmation email (with tests)

## QA Notes
This is totally API based, and will be tested by the front-end team while developing the new user settings pages!

## Documentation

API docs for managing emails aren't there yet, and I'm not sure if we want to make it a user-facing feature?

## Side Effects
None anticipated

## Ticket
https://openscience.atlassian.net/browse/PLAT-1125
https://openscience.atlassian.net/browse/PLAT-1150

### CODE REVIEW

**Comment by aaxelb** [comment]
I apologize in advance for my pedantry.

Based on the meaning of the words, it seems like "confirmed" and "verified" are flipped.

Verification is about identity and truth. "Does this email belong to the user?" The add/merge request is verified when the user clicks the link in the email.

Confirmation is about making a final decision, reasserting the earlier request. "Is the user sure they want this to happen?" The request is confirmed when the user clicks the button in the confirmation modal and the email is added or the account is merged.

buuuut language is chaos, it's probably confusing either way, and as long as it's consistent and clearly documented it'll work just fine \*shrug*

**Comment by aaxelb** [comment]
Oh, more importantly (and less quibbly): We will need to know whether it's an account merge or just adding a new email, since there's different language in the modal. Maybe yet another boolean field?

**Comment by caseyrollins** [comment]
@aaxelb, while I agree that our usage of the words confirmed and verified do not adhere to their true meaning, I'm not sure that reversing confirmed and verified will cause less confusion. 

IMO, the way we're using confirmed here is most consistent with the rest of codebase. The `date_confirmed` field and `is_confirmed` property on the user model reference an account being verified from clicking an email link. The v1 route for verifying emails is also `/confirm/` and we reference users that have not clicked the link in their email as "unconfirmed" users internally (and throughout the codebase). Also, Erin has made the help text for each field very clear in the serializer so I would argue that this approach is clear and consistent, though not semantically correct.

File: api_tests/users/views/test_user_settings.py [hunk 1]
```diff
@@ -245,27 +255,29 @@ def payload(self, user_one):
             }
         }
 
-    def test_get_emails_current_user_no_unconfirmed(self, app, url, user_one):
+    def test_get_emails_current_user(self, app, url, user_one):
         res = app.get(url, auth=user_one.auth)
         assert res.status_code == 200
-        assert len(res.json['data']) == user_one.emails.count()
+        assert len(res.json['data']) == user_one.emails.count() + len(user_one.unconfirmed_emails)
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api_tests/users/views/test_user_settings.py]: Yes, I covered those cases more explicitly in the `TestUserEmailDetail` class, as I thought it'd be easier to explicitly test certain conditions down there. Doesn't mean I can't test the right numbers of each are returned tho, I'll do that!

File: api_tests/users/views/test_user_settings.py [hunk 2]
```diff
@@ -328,6 +340,43 @@ def test_unconfirmed_email_with_expired_token_not_in_results(self, app, url, pay
         assert new_token in returned_ids
         assert old_token not in returned_ids
 
+    def test_filter_by_attributes(self, app, url, user_one):
+        confirmed_not_verified = 'notyet@unconfirmed.test'
+        token = user_one.add_unconfirmed_email(confirmed_not_verified)
+        user_one.email_verifications[token]['confirmed'] = True
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api_tests/users/views/test_user_settings.py]: If I understand your question, I am pretty sure it won't break if another email is added to user one! Here's what an example user's email verification looks like:

    {    
        'AxDkSBGA9F9aBd85fhKoHoSiLFQFSF': {
            'confirmed': False, 
            'external_identity': None, 
            'email': 'save@thewhales.test', 
            'expiration': datetime.datetime(2018, ...)
       }, 'QrOFptHB3lupaoc8aRRh2k9NMTAgR4': {
           'confirmed': True, 
           'external_identity': None, 
           'email': 'notyet@unconfirmed.test', 
           'expiration': datetime.datetime(2018, ..)
        }
    }

The email verifications seem to be stored by token, and so adding another unconfirmed email will add a new token to the dict! Just tried it locally and verified that the test didn't break and that indeed there is one email per token.

[the remark this answers is not in the document]
Comment (caseyrollins) [reply api_tests/users/views/test_user_settings.py]: Yes, that is what I was wondering, thank you!

File: api/users/serializers.py [hunk 3]
```diff
@@ -510,14 +521,26 @@ def create(self, validated_data):
...
         if primary and instance.confirmed:
             user.username = instance.address
             user.save()
         elif primary and not instance.confirmed:
             raise exceptions.ValidationError('You cannot set an unconfirmed email address as your primary email address.')
+
+        if verified and not instance.verified:
+            if not instance.confirmed:
+                raise exceptions.ValidationError('You cannot verify an email address that has not been confirmed by a user.')
+            user.confirm_email(token=instance.id)
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/users/serializers.py]: AAAH wow yes, thank you, these are both excellent calls.

**Comment by jamescdavis** [comment]
The `resend_confirmation` parts of this PR LGTM! :+1:

**Comment by caseyrollins** [comment]
@erinspace, can you bring this branch up to date when you get a chance?

**Pull request merged**